homeassistant/helpers/script.py

```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
-
-            elif action == ACTION_DEVICE_AUTOMATION:
-                referenced.add(step[CONF_DEVICE_ID])
-
-        self._referenced_devices = referenced
-        return referenced
+    def _config_cache(self):
+        return self._script._config_cache  # pylint: disable=protected-access
 
     @property
-    def referenced_entities(self):
-        """Return a set of referenced entities."""
-        if self._referenced_entities is not None:
-            return self._referenced_entities
-
-        referenced = set()
-
-        for step in self.sequence:
-            action = _determine_action(step)
-
-            if action == ACTION_CALL_SERVICE:
-                data = step.get(service.CONF_SERVICE_DATA)
-                if not data:
-                    continue
-
-                entity_ids = data.get(ATTR_ENTITY_ID)
+    def name(self):
+        """Return script name."""
+        return self._script.name
 
-                if entity_ids is None:
-                    continue
-
-                if isinstance(entity_ids, str):
-                    entity_ids = [entity_ids]
-
-                for entity_id in entity_ids:
-                    referenced.add(entity_id)
-
-            elif action == ACTION_CHECK_CONDITION:
-                referenced |= condition.async_extract_entities(step)
+    @property
+    def last_action(self):
+        """Return last action."""
+        return self._last_action
 
-            elif action == ACTION_ACTIVATE_SCENE:
-                referenced.add(step[CONF_SCENE])
+    @last_action.setter
+    def last_action(self, value):
+        self._last_action = value
 
-        self._referenced_entities = referenced
-        return referenced
-
-    def run(self, variables=None, context=None):
+    @abstractmethod
+    async def async_run(self) -> None:
         """Run script."""
-        asyncio.run_coroutine_threadsafe(
-            self.async_run(variables, context), self.hass.loop
-        ).result()
-
-    async def async_run(
-        self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
 
-        This method is a coroutine.
-        """
-        self.last_triggered = date_util.utcnow()
-        if self._cur == -1:
-            self._log("Running script")
-            self._cur = 0
-
-        # Unregister callback if we were in a delay or wait but turn on is
-        # called again. In that case we just continue execution.
-        self._async_remove_listener()
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
+    async def _async_run(self, start):
+        for self._step, self._action in islice(
+            enumerate(self._script.sequence), start, None
+        ):
             try:
-                await self._handle_action(action, variables, context)
+                await getattr(self, f"_async_{_determine_action(self._action)}_step")()
             except _SuspendScript:
```

Comment:
No need to catch it if you're just raising it.

Reply:
Actually, there is, because otherwise it gets caught by the `except Exception` clause that follows.